README.md
## Todo App React

A todo app with a React frontend and Express.js backend. It uses the Firebase real-time database and is planned to be hosted on Heroku. 

This is used in conjunction with an [iOS todo app](https://github.com/muhdmirzamz/TodoApp) I'm working on. Both of these projects connect to the same Firebase database.


### INSTALL
This project requires that you setup your own Firebase database. If you want to run this project, you need your own config file from Firebase. Once done, you can continue with the installation.

`npm install`  
`npm start`


### Technologies used:
- React
- Express.js
- React routing
- Firebase
- Heroku


CHANGELOG.md
### Changelog

14 April 2023:
- The todos list now refreshes to the latest todos when you add a new todo to the list

4 April 2023:
- You can now get the todos for the first todo list on initial load 

2 April 2023:
- You can now get todos after selecting the dropdown beside "Choose a list"

8 March 2021:
- You can now create new lists
- You can now add to different lists

7 March 2021:
- There is now a user interface where you can select lists

28 February 2021:
- You can now create lists
- Fetching lists is done
- Changed database structure

23 February 2021:
- Users can now sign out

22 February 2021:
- Added a GET request to check for missing API key

5 February 2021:
- Updated dependency versions

5 October 2020:
- Added a delete button

29 September 2020:
- Register page redirects to Login page after successful operation
- You can now delete an item by clicking on the item

12 September 2020:
- You can add items to the list and the list will update

5 September 2020:
- Registration mechanism completed
- You can add items to the database
- You can retrieve items from the database
- Fixed: Alerts and popups now work

9 August 2020:
- Login mechanism complete

25 July 2020:
- Login mechanism 90% complete
- Removed history package

23 July 2020:
- Added a form
- Removed Firebase credentials
- Implemented routing

17 July 2020:
- Able to get POST data from frontend

16 July 2020:
- Initial commit


backend/app.js
const express = require('express')

// imported as documented from Firebase
const firebase = require('firebase/app')
require('firebase/auth')
require('firebase/database')

// var bodyParser = require('body-parser')

const app = express()

app.use(express.urlencoded({ extended: true }));
app.use(express.json());

const port = 9000



var firebaseConfig = {
    apiKey: "AIzaSyCG6_Oh8dOzt21flBDelf0eh4sRq9Fuo1k",
  authDomain: "todoapp-id.firebaseapp.com",
  databaseURL: "https://todoapp-id.firebaseio.com",
  projectId: "todoapp-id",
  storageBucket: "todoapp-id.appspot.com",
  messagingSenderId: "953077094877",
  appId: "1:953077094877:web:3a4d99d569a2c4655dfe34"
};

firebase.initializeApp(firebaseConfig)


app.get("/test", (req, res) => {
    res.send('Hello World')
});

app.get("/check", (req, res) => {
    if (Object.keys(firebaseConfig).length === 0) {
        // this appears in the client's catch block
        res.status(500).send('Dont forget your API key');
    }
});

app.post("/register", (req, res) => {
    console.log('/register Got body:', req.body);

    firebase.auth().createUserWithEmailAndPassword(req.body.email, req.body.password).catch((error) => {
        var errorCode = error.code;
        var errorMessage = error.message;

        console.log(errorCode + " " + errorMessage)
    })

    res.send('Got a post')
})



app.post("/signin", (req, res) => {
    console.log('/signin Got body:', req.body);

    firebase.auth().signInWithEmailAndPassword(req.body.email, req.body.password).catch((error) => {
        if (error) {
            console.log(error.message)
        }
    })


    firebase.auth().onAuthStateChanged((user) => {
        if (user) {
          // User is signed in.
          console.log(user.email)

          res.send('hello there')
        } else {
          // No user is signed in.
        }
    });
})

app.post("/addList", (req, res) => {
    console.log("Got body: " + req.body)

    // might wanna add a check for valid user here
    var user = firebase.auth().currentUser

    var database = firebase.database();

    // get the new key
    const newItemRef = database.ref().child("lists").child(user.uid).push()

    // set the new item
    newItemRef.set({
        listName: req.body.listName
    });

    // push the new key so that frontend will be able to append to new object
    res.status(200).send(newItemRef.path.pieces_[1])
})


// you need to send over the userId
app.get("/getLists", (req, res) => {
    console.log("getting lists")

    var userId = firebase.auth().currentUser.uid;
    
    firebase.database().ref("lists/" + userId).once('value').then(function(snapshot) {
        console.log("Got lists")
        console.log(snapshot.val())

        res.send(snapshot.val())
    });
})

// you need to send over the listId
app.get("/getTodos", (req, res) => {
    console.log("getting todos")

    //let listId = firebase.auth().currentUser.uid;
    
    let listId = req.query.listId

    firebase.database().ref("todos/" + listId).once('value').then(function(snapshot) {
        console.log("Got todos")
        console.log(snapshot.val())

        res.send(snapshot.val())
    });
})

// you need to send over 2 properties:
// listId
// todoItem
app.post("/addTodo", (req, res) => {
    console.log("ADD TODO==== Got body: " + req.body)

    var database = firebase.database();

    var listId = req.body.listId
    var timestamp = req.body.timestamp

    // get the new key
    const newItemRef = database.ref().child("todos").child(listId).push()
    var key = newItemRef.key
    console.log(`key: ${key}`)

    // set the new item
    var newItem = {
        todoItem: req.body.todoItem,
        timestamp: timestamp
    }

    newItemRef.set(newItem);

    // push the new key so that frontend will be able to append to new object
    res.status(200).send(key) //newItemRef.path.pieces_[1]
})

app.delete("/deleteItem", (req, res) => {
    console.log("Got body: ")

    // https://github.com/axios/axios/issues/736#issuecomment-307209067
    console.log(req.query.item)

    // might wanna add a check for valid user here
    var user = firebase.auth().currentUser

    var database = firebase.database();

    // get the new key
    const newItemRef = database.ref().child(user.uid).child(req.query.item).remove()

    res.send('Got a DELETE')
})

app.get("/signout", (req, res) => {
    firebase.auth().signOut().then(() => {
        // Sign-out successful.
        console.log("Sign out successful")

        res.send('SUCCESS')
    }).catch((error) => {
        // An error happened.
    });
})

app.listen(port, () => console.log("App listening at https://localhost:${" + port + "}"))

frontend/src/index.js
import React from 'react';
import ReactDOM from 'react-dom';

import './index.css';

import Dashboard from './dashboard'
import Register from './register'

import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link
} from 'react-router-dom'

import axios from "axios"

import * as serviceWorker from './serviceWorker';

class Page extends React.Component {
  constructor(props) {
    super(props)
  }

  render() {
    return(
      <Router>
        <Switch>
          <Route exact path='/' component={Home} />
          <Route path='/dashboard' component={Dashboard} />
          <Route path='/register' component={Register} />
        </Switch>
      </Router>
    )
  }
}

class Home extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      email: '', 
      password: ''
    }

    this.onChangeEmailText = this.onChangeEmailText.bind(this)
    this.onChangePasswordText = this.onChangePasswordText.bind(this)

    this.onSubmit = this.onSubmit.bind(this)
  }

  componentDidMount() {
    axios.get("/check").catch((error) => {
      if (error) {
          alert(error.response.data)
      }
    })
  }

  onChangeEmailText(event) {
    this.setState({email: event.target.value})
  }

  onChangePasswordText(event) {
    this.setState({password: event.target.value})
  }

  onSubmit(event) {
    // axios.get("/test").then((response) => {
    //   console.log(response)
    // })

    axios.post("/signin", {email: this.state.email, password: this.state.password}).then((response) => {
      if (response.status === 200) {
        console.log(response)

        alert('Sign in successfully')

        // https://stackoverflow.com/a/53916596
        this.props.history.push('/dashboard')
      }
    }).catch((error) => {
      if (error) {
          console.log(error)
          console.log(error.message)
          console.log(error.response)
          
      }
    })

    event.preventDefault()
  }

  render() {
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          <label>Email: </label>
          <input type='text' value={this.state.email} onChange={this.onChangeEmailText} />

          <label>Password: </label>
          <input type='password' value={this.state.password} onChange={this.onChangePasswordText} />

          <input type='submit' value='Sign in' />
        </form>
        <Link to="/register">Register</Link>
      </div>
    );
  }
}

ReactDOM.render(
  <Page />,
  document.getElementById('root')
);

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: https://bit.ly/CRA-PWA
serviceWorker.unregister();


frontend/src/register.js
import React from 'react';
import ReactDOM from 'react-dom';

import axios from 'axios';

class Register extends React.Component {
    constructor(props) {
        super(props)

        this.state = {username: '', password: ''}

        this.onChangeUsernameText = this.onChangeUsernameText.bind(this)
        this.onChangePasswordText = this.onChangePasswordText.bind(this)

        this.onSubmit = this.onSubmit.bind(this)
    }

    onChangeUsernameText(event) {
        this.setState({username: event.target.value})
    }

    onChangePasswordText(event) {
        this.setState({password: event.target.value})
    }

    onSubmit(event) {
        axios.post("/register", {email: this.state.username, password: this.state.password}).then((response) => {
            if (response.status === 200) {
                console.log(response)

                alert('Registered successfully')

                this.props.history.push('/')
            } else {
                alert('Oops something went wrong!')
            }
        })

        event.preventDefault()
    }

    render() {
        return(
            <form onSubmit={this.onSubmit}>
                <label>Username: </label>
                <input type='text' value={this.state.username} onChange={this.onChangeUsernameText} />

                <label>Password: </label>
                <input type='password' value={this.state.password} onChange={this.onChangePasswordText} />

                <input type='submit' value='Submit' />
            </form>
        )
    }
}

export default Register;

frontend/src/dashboard.js
import React from 'react'

import axios from 'axios'
import { format } from 'date-fns'

class Dashboard extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      listFieldText: '', 
      todoFieldText: '', 
      lists: {}, 
      todos: {}, 
      selectedListId: ''
    }

    this.onChangeListFieldText = this.onChangeListFieldText.bind(this)
    this.onChangeTodoFieldText = this.onChangeTodoFieldText.bind(this)
    this.onClickListItem = this.onClickListItem.bind(this)

    this.onListSubmit = this.onListSubmit.bind(this)
    this.onTodoSubmit = this.onTodoSubmit.bind(this)
    
    this.onDelete = this.onDelete.bind(this)

    this.onSignout = this.onSignout.bind(this)
  }

  componentDidMount() {
    axios.get("/getLists").then((response) => {
      console.log("Get items status: " + response.status)

      if (response.status === 200) {
        // returns as an object that is made up of a list of key (key is the list id) + value (value is the list name - being an object)
        console.log(`lists: ${response.data}`)

        let todoListIds = Object.keys(response.data)         

        // just make sure we have something in the array
        if (todoListIds.length > 0) {

          // set the list id to start off at the first element
          var id = todoListIds[0]
          console.log("Current list id: " + id)

          this.setState({lists: response.data, selectedListId: id})



          axios.get(`/getTodos`, {
            params: {
              listId: id
            }
          }).then(response => {
            if (response.status === 200) {
              console.log(response.data)
      
              // returns as an object that is made up of a list of key + value (value being an object)
              console.log(`todos: ${JSON.stringify(response.data)}`)
      
              this.setState({todos: response.data})

            } else {
              alert("Oops something went wrong!")
            }
          })
        }
      } else {
        console.log("response is not 200")
      }
    })
  }

  onChangeListFieldText(event) {
    this.setState({listFieldText: event.target.value})
  }

  onChangeTodoFieldText(event) {
    this.setState({todoFieldText: event.target.value})
  }

  onClickListItem(event) {

    // get the selected index
    var selectedIndex = document.getElementById("lists").selectedIndex;

    
    // get the key at the selected index
    // our list has the key inside but it does not have a "property" attached to it 
    // it's just {kjvskjvnvla: listName: 'test'}

    // var keyAtSelectedIndex = Object.keys(this.state.lists)[selectedIndex]

    // get the listName from the list object (you are accessing an object's key)
    
    // var listName = this.state.lists[keyAtSelectedIndex].listName


    // selected index == selected key's index
    var listId = Object.keys(this.state.lists)[selectedIndex]

    this.setState({selectedListId: listId})


    axios.get(`/getTodos`, {
      params: {
        listId: listId
      }
    }).then(response => {
      if (response.status === 200) {
        console.log(response.data)

        // returns as an object that is made up of a list of key + value (value being an object)
        console.log(`todos: ${JSON.stringify(response.data)}`)

        this.setState({todos: response.data})
        // this.setState({todos: response.data})



        // var newLists = {...this.state.lists}
        // newLists[response.data] = {listName: this.state.listFieldText}

        // spread operator syntax to copy an object and add an item to it
        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax

        // const newTodos = {
        //   ...this.state.todos,
        //   [response.data]: {todoItem: this.state.todoFieldText}
        // }

        // this.setState({lists: newLists, listFieldText: ''})
      } else {
        alert("Oops something went wrong!")
      }
    })

  }

  onListSubmit(event) {
    axios.post("/addList", {listName: this.state.listFieldText}).then((response) => {
      console.log("Add items status: " + response.status)

      if (response.status === 200) {
        // new key is in response.data
        alert(response.data)

        var newLists = {...this.state.lists}
        newLists[response.data] = {listName: this.state.listFieldText}

        // spread operator syntax to copy an object and add an item to it
        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax

        // const newTodos = {
        //   ...this.state.todos,
        //   [response.data]: {todoItem: this.state.todoFieldText}
        // }

        this.setState({lists: newLists, listFieldText: ''})
      } else {
        alert("Oops something went wrong!")
      }
    })

    event.preventDefault()
  }


  onTodoSubmit(event) {
    var timestamp = format(new Date(), 'yyyy-MM-dd HH:mm:ss.sss')

    axios.post("/addTodo", {todoItem: this.state.todoFieldText, listId: this.state.selectedListId, timestamp: timestamp}).then((response) => {
      console.log("Add items status: " + response.status)

      if (response.status === 200) {
        // new key is in response.data
        alert(response.data)

        var newTodos = {...this.state.todos}
        newTodos[response.data] = {timestamp: timestamp, todoItem: this.state.todoFieldText}


        this.setState({todos: newTodos, todoFieldText: ''})

        // spread operator syntax to copy an object and add an item to it
        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax

        // const newTodos = {
        //   ...this.state.todos,
        //   [response.data]: {todoItem: this.state.todoFieldText}
        // }

        // this.setState({lists: newLists, listFieldText: ''})
      } else {
        alert("Oops something went wrong!")
      }
    })

    event.preventDefault()
  }

  

  onDelete(key) {
    alert("Deleted: " + key)

    var obj = this.state.todos
    delete obj[key] // delete a key from an object
    this.setState({todos: obj})


    // needs to specify an extra 'params' key
    // read the whole issue/thread
    // https://github.com/axios/axios/issues/736
    // https://github.com/axios/axios/issues/736#issuecomment-307209067
    axios.delete("/deleteItem", {params: {item: key}}).then((response) => {
      console.log("Add items status: " + response.status)

      if (response.status === 200) {
        console.log(response.data)
      }
    })
  }

  onSignout(event) {
    axios.get("/signout").then((response) => {
      if (response.status === 200) {
        console.log(response.data)

        alert('Sign out successfully')

        this.props.history.push('/')
      }
    })
  }

  render() {
    return(
      <div>
        {/* Add list */}
          <form onSubmit={this.onListSubmit}>
            <label>Add list: </label>
            <input type='text' value={this.state.listFieldText} onChange={this.onChangeListFieldText} />

            <input type='submit' value='Submit' />
          </form>


          {/* display list */}
          Lists:
          <ul>
            {
              Object.keys(this.state.lists).map((key) => {
                // using a sub component to avoid re-render
                // https://stackoverflow.com/a/29810951
                return <ListItem key={key} itemKey={key} deleteFunction={this.onDelete} value={this.state.lists[key].listName} />
              })
            }
          </ul>

          {/* Choose list */}
          <label htmlFor="lists">Choose a list:</label>

          <select name="lists" id="lists" onChange={this.onClickListItem}>
            {
              Object.keys(this.state.lists).map((key) => {
                return <option key={key} value={this.state.lists[key].listName}> {this.state.lists[key].listName} </option>
              })
            }
          </select>



          {/* Add todo */}
          <form onSubmit={this.onTodoSubmit}>
            <label>Add todo: </label>
            <input type='text' value={this.state.todoFieldText} onChange={this.onChangeTodoFieldText} />

            <input type='submit' value='Submit' />
          </form>



          {/* <ul>
            {
              Object.keys(this.state.lists).map((key) => {
                // using a sub component to avoid re-render
                // https://stackoverflow.com/a/29810951
                return <ListItem key={key} itemKey={key} deleteFunction={this.onDelete} value={this.state.lists[key].listName} />
              })
            }
          </ul> */}

          <ul>
            {
              Object.keys(this.state.todos).map((key) => {
                // using a sub component to avoid re-render
                // https://stackoverflow.com/a/29810951
                return <ListItem key={key} itemKey={key} deleteFunction={this.onDelete} value={this.state.todos[key].todoItem} />
              })
            }
          </ul>

          <input type='button' onClick={this.onSignout} value='Sign out' />
      </div>
    )
  }
}

class ListItem extends React.Component {

  constructor(props) {
    super(props)
  }

  render() {

    // https://stackoverflow.com/questions/39226757/react-passing-parameter-with-arrow-function-in-child-component
    // you are not passing down the argument through the prop itself
    // you are passing down the function and the argument as the prop
    return (
      <div>
        <li>{this.props.value}</li>
        <button onClick={() => this.props.deleteFunction(this.props.itemKey)}>Delete</button>
      </div>
    )
  }
}


export default Dashboard;